llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp

```diff
@@ -1035,21 +1035,40 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
     }
     break;
   }
-  case Intrinsic::fabs:
+  case Intrinsic::fabs: {
+    auto LT = getTypeLegalizationCost(RetTy);
+    if (ST->hasVInstructions() && LT.second.isVector()) {
+      // lui a0, 8
+      // addi a0, a0, -1
+      // vsetvli a1, zero, e16, m1, ta, ma
+      // vand.vx v8, v8, a0
+      // f16 with zvfhmin and bf16 with zvfhbmin
+      if (LT.second.getVectorElementType() == MVT::bf16 ||
+          (LT.second.getVectorElementType() == MVT::f16 &&
+           !ST->hasVInstructionsF16()))
+        return LT.first * getRISCVInstructionCost(RISCV::VAND_VX, LT.second,
+                                                  CostKind) +
+               2;
+      else
+        return LT.first *
+               getRISCVInstructionCost(RISCV::VFSGNJX_VV, LT.second, CostKind);
+    }
+    break;
+  }
   case Intrinsic::sqrt: {
     auto LT = getTypeLegalizationCost(RetTy);
-    // TODO: add f16/bf16, bf16 with zvfbfmin && f16 with zvfhmin
     if (ST->hasVInstructions() && LT.second.isVector()) {
-      unsigned Op;
-      switch (ICA.getID()) {
-      case Intrinsic::fabs:
-        Op = RISCV::VFSGNJX_VV;
-        break;
-      case Intrinsic::sqrt:
-        Op = RISCV::VFSQRT_V;
-        break;
-      }
-      return LT.first * getRISCVInstructionCost(Op, LT.second, CostKind);
+      SmallVector<unsigned, 3> Opcodes;
+      // f16 with zvfhmin and bf16 with zvfbfmin
+      if (LT.second.getVectorElementType() == MVT::bf16)
```

Comment:
Do we need to account for <vscale x 32 x f16> and <vscale x 32 x bf16> needing to be split before we can widen it? The widening of those types would produce an LMUL=16 value so we have to split first.

Reply:
1. It seems is legal for <vscale x 32 x f16> and <vscale x 32 x bf16>,  at now LT.first =1, but from asssemble really spilt . I'm not sure how to set the conditions for spilt, because in theory this scenario is also a split
![image](https://github.com/user-attachments/assets/e0ec3555-4ff0-47cf-b714-5688c41767b4)

Reply:
cost of vscale x 16 x f16 is 12, whther cost of vscale x 32 x f16 is 24 and  vscale x 64 x f16 is 36????

Reply:
formula： 
1. Because the SpiltOpcodes ==  Opcodes,  (LT.first - 1 ) * getRISCVInstructionCost(SpiltOpcodes  , LT.second, CostKind) + getRISCVInstructionCost(Opcodes, LT.second, CostKind)  ----> LT.first * getRISCVInstructionCost(Opcodes, LT.second, CostKind)

Reply:
The real issue is that the vfsqrt cost needs to use an f32 type and that type needs to be split if it isn't legal.

Reply:
It has been custom and has not used f32 type？you means  cost of vscale x 32 x f16 == cost of vscale x 32 x f32??
![image](https://github.com/user-attachments/assets/b899c9fa-2501-446d-871f-e57aed9f2f56)

Reply:
I don't think 12 is the correct cost for <vscale x 16 x f16>. That cost did not use f32 for the vsqrt.v cost.

The cost for <vscale x 4 x bfloat> should be at least 4 and possibly 5 or 6, not 3. There's a vfwcvt.f.f.v from EEW=16 EMUL=m1 to EEW=32 EMUL=m2. That's either a cost of 1 if the hardware can produce an EMUL=2 result in 1 cycle or its 2 if it takes 2 cycles to produce a wide result.

Then there's a EEW=32 EMUL=2 vfsqrt.v. That should be a cost of 2 since the EMUL is 2.

Then there's a vfncvt.f.f.w from EEW=32 EMUL=2 to EEW=16 EMUL=1. That's either 1 cycle if the hardware can narrow EMUL=2 in 1 cycle. Or it's a cost of 2.

So the total cost is either (1 + 2 + 1) or (2 + 2 + 1) or (1 + 2 + 2) or (2 + 2 + 2).